### PR TITLE
bpo-33625: Release GIL for grp.getgr{nam,gid} and pwd.getpw{nam,uid}

### PR DESCRIPTION
**Pull request by william-gr**
Especially when using more complex nss modules the call might that an
unknown amount of time depending on the service in question.
This makes sures others threads are not blocked waiting the call to
finish.

https://bugs.python.org/issue33625

### CODE REVIEW

**Comment by tiran** [comment]
LGTM, thanks!

Please use the blurb tool to add a news entry.

**Comment by william-gr** [comment]
@tiran NEWS added, let me know if thats not to your liking, its my first time doing this ; )

Thanks!

File: Misc/NEWS.d/next/Library/2018-05-23-17-07-54.bpo-33625.nzQgD8.rst
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2018-05-23-17-07-54.bpo-33625.nzQgD8.rst]: Sphinx complains about using the default role. Write as ``` :func:`pwd.getpwnam` ``` or ``` ``pwd.getpwnam()`` ```.

**Review by serhiy-storchaka — COMMENTED** [review]
What about `getpwall()`? And is it worth to release GIL in the `grp` and the `spwd` modules?

**Comment by serhiy-storchaka** [comment]
You are right, and switching a thread between `setpwent()` and `getpwent()` is not safe. We would need to add a special global lock for `getpwall()`. Yet one example when GIL makes the code safer and simpler.

**Comment by tiran** [comment]
Can you either fix ``getgrgid`` and ``getgrnam`` in this PR or open up another PR. Don't bother about ``spwd``. The module shouldn't be used any more.

**Comment by william-gr** [comment]
@serhiy-storchaka You are absolutely right about re-entrant functions. However I am not sure whats the best way to check if they are available in the system in python project. Should I add a check in configure.ac for the existence of each of them? HAVE_GETPWNAM_R, HAVE_GETPWUID_R ?

**Comment by william-gr** [comment]
I am now handling ERANGE of _r functions, will update patch soon.

**Comment by william-gr** [comment]
Updated handling ERANGE, let me know how that looks and/or any suggestions/changes.

Thank you again.

**Comment by william-gr** [comment]
Windows-PR seems to be failing but I dont know why, I cant find logs in the link. How should I proceed?

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: This is a constant size buffer. It is better to use a local variable for it.
```c
struct group grp;
...
status = getgrgid_r(gid, &grp, buf, bufsize, &p);
```

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: Use `PyMem_RawMalloc()`.

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: Please add a space between `if` and `(` for conforming PEP 7.

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: Use `p` instead of `grpbuf`.

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: This condition is doubled in the `while` below. It would be better to write:
```c
if (grpbuf != NULL || status != ERANGE) {
    break;
}
```
and use `while (1)`.

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: It could be more efficient to use realloc.

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: What is this check for?

Comment (william-gr) [reply Modules/grpmodule.c]: This is to make sure we are not allocating too big of a buffer (its a maximum cap)

**Comment by william-gr** [comment]
@serhiy-storchaka thanks for the review. Let me know if I have addressed them to your liking.

**Comment by william-gr** [comment]
Is there a way I can trigger a new macOS build? The failure does not seem related to the commit.

**Comment by william-gr** [comment]
Ping. Wondering if there is anything else to do before it can be reviewed or just wait. Thanks

**Comment by ned-deily** [comment]
As mentioned on the issue tracker, I do not think we should backport the updated PR to 3.6 or 2.7 due to the much-more extensive changes.   If we act fast, it could still be a candidate for 3.7.0rc1.

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: The official documentation recommends to use `sysconf(_SC_GETGR_R_SIZE_MAX)`. The comment near the `NSS_BUFLEN_GROUP` definition in Linux headers suggests the same.

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: This limit looks arbitrary. Use `if (bufsize > PY_SSIZE_T_MAX)` and set MemoryError by `PyErr_NoMemory()`.

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: `PyMem_RawRealloc()` can fail. In this case it will return NULL.
```c
p = PyMem_RawRealloc(buf, bufsize);
if (p == NULL) {
    PyErr_NoMemory();
    break;
}
buf = p;
```

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: If the loop is infinite, it is better to write it as `while (1) { ... }` rather of `do { ... } while (1);`.

**Review by serhiy-storchaka — COMMENTED** [review]
Look also at examples in the official Posix documentation: http://pubs.opengroup.org/onlinepubs/9699919799/functions/getgrgid.html.

File: Modules/grpmodule.c
Comment (vstinner) [inline Modules/grpmodule.c]: You leak buf memory here.

File: Modules/grpmodule.c
Comment (vstinner) [inline Modules/grpmodule.c]: You leak buf here.

File: configure.ac
Comment (vstinner) [inline configure.ac]: Please don't hardcode a value here. IMHO it's better to handle that the define is missing in Modules/pwdmodule.c.

File: configure.ac
Comment (vstinner) [inline configure.ac]: autoconf generates this for you, no?

File: configure.ac
Comment (vstinner) [inline configure.ac]: Why not just adding the check to the existing long "# checks for library functions
AC_CHECK_FUNCS(alarm accept4 setitimer getitimer bind_textdomain_codeset chown ..." list?

File: Modules/grpmodule.c
Comment (william-gr) [reply Modules/grpmodule.c]: I am failing to see how. If p == NULL it will free buf in the if statement before.

File: Modules/grpmodule.c
Comment (william-gr) [reply Modules/grpmodule.c]: I am failing to see how. If p == NULL it will free buf in the if statement before.

Comment (vstinner) [reply Modules/grpmodule.c]: Oh. I was confused because one RawFree in done when the GIL is released, and the other one is done when we hold the GIL. I would suggest to release buf when we hold the GIL, and change the previous if as:

```
    if (status != 0) {
        p = NULL;
    }
```

File: configure.ac
Comment (william-gr) [reply configure.ac]: Yes, sorry, I am not too familiar with autotools, will change to use that instead, thanks for the pointer.

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: Sorry, it should be `bufsize > PY_SSIZE_T_MAX / 2` or `bufsize > (PY_SSIZE_T_MAX >> 1)`.

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: The MemoryError will be overridden by a KeyError below. You need to avoid this. For example use a special flag or check `PyErr_Occurred()` before setting a KeyError.

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: Is it possible that p != NULL if status != 0?

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [inline Modules/grpmodule.c]: `sysconf(_SC_GETGR_R_SIZE_MAX)` can return -1. Check this and use some reasonable default in this case.

The type of bufsize should be Py_ssize_t, or the test for integer overflow below should be more complex.

File: Modules/grpmodule.c
Comment (william-gr) [reply Modules/grpmodule.c]: Yes, p may be set to some arbitrary entry on some error conditions of getgrgid_r and similar methods.

**Comment by william-gr** [comment]
Pushed another update, let me know if thats good enough!

**Comment by william-gr** [comment]
Windows PR failing, I suppose thats just because the build is fragile? Last time it was macOS that failed

**Comment by vstinner** [comment]
> Windows PR failing, I suppose thats just because the build is fragile? Last time it was macOS that failed

I fixed many bugs in master, like the test_asyncio bug that you had on VSTS. Please merge master into your PR to get latest fixes.

**Comment by william-gr** [comment]
Got it, just rebased from master, thanks for getting back to me.

File: Modules/grpmodule.c
Comment (vadmium) [inline Modules/grpmodule.c]: I don’t see the point of calling PyErr_NoMemory when you set RuntimeError later on.

Also, setting _status_ seems redundant because you only get here if you get ERANGE.

Comment (william-gr) [reply Modules/grpmodule.c]: You're right about ERANGE, will fix.

@serhiy-storchaka you proposed the PyErr_NoMemory, how should I properly raise that error? I used RuntimeError afterwards but that does not seem right.

Thank you

Comment (serhiy-storchaka) [reply Modules/grpmodule.c]: Wouldn't MemoryError be more appropriate that RuntimeError?

Comment (vstinner) [reply Modules/grpmodule.c]: We use MemoryError on integer overflow when allocating memory. We do it all the time in unicodeobject.c for example.

**Comment by william-gr** [comment]
Updated to use PyErr_NoMemory

**Comment by william-gr** [comment]
Did we miss the deadline to get this in time for 3.7?

**Comment by ned-deily** [comment]
For 3.7.0, yes, sorry. But let’s get it into master and exposed there.  We can then consider whether it might be appropriate for a backport to a 3.7.x maintainence release as a bug fix.

**Comment by vstinner** [comment]
> Did we miss the deadline to get this in time for 3.7?

It's too late for 3.7.0.

@tiran considers that it's a bugfix, and so should be fixed in 2.7, 3.6, 3.7 and master:
https://bugs.python.org/issue33625#msg317447

So it can be fixed in 3.7.1.

**Comment by ned-deily** [comment]
> So it can be fixed in 3.7.1.

See the discussion on the bug tracker about this.  I think we should hold off on backporting to 3.7 for the moment.  Let's get it into master first.

**Comment by william-gr** [comment]
Can this be merged to master, then? Or what is blocking?

**Comment by william-gr** [comment]
Ping?

This has been an issue on threaded python software on systems connected to large AD/LDAP environments.

If this solution is not OK can any of you point me to a better solution, please?

File: Misc/NEWS.d/next/Library/2018-05-23-17-07-54.bpo-33625.nzQgD8.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2018-05-23-17-07-54.bpo-33625.nzQgD8.rst]: The NEWS entry is not correct. The GIL is not *always* released. It depends if reentrant variants of these functions are available.

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by william-gr** [comment]
I have made the requested changes; please review again

Thank you

File: Modules/grpmodule.c
Comment (vstinner) [inline Modules/grpmodule.c]: PyMem_RawFree(NULL) is valid (does nothing), the if() is useless.

File: Modules/grpmodule.c
Comment (vstinner) [inline Modules/grpmodule.c]: It's strange to reuse the "struct group *p;" variable to reallocate the "char *buf = NULL;" variable: C types are not the same!? I wold prefer to see a different variable (ex: "buf2").

File: Modules/grpmodule.c
Comment (vstinner) [inline Modules/grpmodule.c]: Would you mind to add a "#define DEFAULT_BUFFER_SIZE 1024" at top level, rather than using an hardcoded constant here?

File: Modules/grpmodule.c
Comment (vstinner) [inline Modules/grpmodule.c]: nitpick: the PEP 7 requires

```
}
else {
```

File: Modules/pwdmodule.c
Comment (vstinner) [inline Modules/pwdmodule.c]: I would prefer to move this just after the get...() call.

**Review by vstinner — CHANGES_REQUESTED** [review]
New review. My comments apply to the all modified functions.

**Comment by william-gr** [comment]
I have made the requested changes; please review again

Thank you

File: Modules/grpmodule.c
Comment (vstinner) [inline Modules/grpmodule.c:248]: Hum, I see a bug which is not related to your change, but while we modify these functions, it would be nice to fix it.

name_chars is encoded to the filesystem encoding, whereas %s decodes it from UTF-8. If the filesystem encoding is not UTF-8, you get mojibake.

The fix is simple: use %S format and pass name: name is always a Unicode string.

File: Modules/pwdmodule.c
Comment (vstinner) [inline Modules/pwdmodule.c:256]: Same comment: please use the %S format here, but if you modify this function, would you mind to also fix the name of the parameter? Rename "arg" to "name", as in Doc/library/pwd.rst. Maybe rename the char* name to name_chars, as in grp_getgrnam_impl().

File: Modules/grpmodule.c
Comment (vstinner) [inline Modules/grpmodule.c]: Hum, wait, you don't check if PyMem_Malloc() failed here? I'm not sure that it's safe to call get*() functions with buf=NULL.

I have a proposition. Move the memory allocation at the start of the loop, and always use PyMem_Realloc().  PyMem_Realloc(buf, bufsize) behaves as PyMem_Malloc(). Pseudo-code:

```
/* PyMem_Malloc code removed from here */
while (1) {
        buf2 = PyMem_RawRealloc(buf, bufsize);
        if (buf2 == NULL) {
            nomem = 1;
            break;
        }
        buf = buf2; /* buf cannot be NULL from this point */
        (...)
    /* no more PyMem_Realloc code here neither */
}
```

(I don't propose to add these comments, it's just to explain my idea ;-))

**Review by vstinner — CHANGES_REQUESTED** [review]
Ah! Honestly, the new PR is way better than the first version! I was going to merge it, when I spotted another issue: you don't check if PyMem_Malloc() fails. IMHO it's not a good idea to call get*() functions with buf=NULL. Maybe the function fails, maybe you get a crash... I would prefer to avoid the risk of crash :-) I proposed to move code which allocates the memory to avoid redundancy and to make sure that get*() are never called with buf=NULL.

By the way, I also proposed to fix a mojibake (encoding) issue while we are on these functions.

Maybe the mojibake issue can be fixed in a separated PR, since I don't think that we are going to backport this one to 3.7 and older. So maybe wait until this PR is merged, and then write a second PR to fix the mojibake issue, and we can easily backport the second PR to all supported branches.

**Comment by william-gr** [comment]
I have made the requested changes; please review again

Thanks for pointing that out and your patience with the reviews. I am glad you pointed the Malloc issue, I also found other problem with latest Realloc change I had made.

I will take your advise and create another PR for mojibake.

**Review by vstinner — APPROVED** [review]

**Comment by vstinner** [comment]
I merged your PR, thanks!

Would you mind to write a second PR to fix the encoding issue that I spootted? I explained how to fix it (use %S format).

File: Modules/grpmodule.c
Comment (serhiy-storchaka) [reply Modules/grpmodule.c:248]: If name is always a Unicode string, you can use %U.

It may be worth to use %R for the case if the name contains invisible characters or trailing whitespaces.

Comment (vstinner) [reply Modules/grpmodule.c:248]: See https://bugs.python.org/issue34604 :-)

**Pull request merged**
